paho/client.go

```diff
@@ -641,133 +655,140 @@ func (c *Client) Subscribe(ctx context.Context, s *Subscribe) (*Suback, error) {
 		}
 	}
 
-	c.debug.Printf("subscribing to %+v", s.Subscriptions)
-
-	subCtx, cf := context.WithTimeout(ctx, c.PacketTimeout)
-	defer cf()
-	cpCtx := &CPContext{subCtx, make(chan packets.ControlPacket, 1)}
+	respInternal := make(chan packets.ControlPacket)
+	resp := make(chan packets.ControlPacket, 1)
 
 	sp := s.Packet()
 
-	mid, err := c.MIDs.Request(cpCtx)
+	var err error
+	sp.PacketID, err = c.MIDs.Request(&CPContext{context.TODO(), respInternal})
 	if err != nil {
 		return nil, err
 	}
-	defer c.MIDs.Free(mid)
-	sp.PacketID = mid
 
-	c.debug.Println("sending SUBSCRIBE")
-	if _, err := sp.WriteTo(c.Conn); err != nil {
-		return nil, err
-	}
-	c.debug.Println("waiting for SUBACK")
-	var sap packets.ControlPacket
+	c.debug.Printf("subscribing to %+v", s.Subscriptions)
 
-	select {
-	case <-subCtx.Done():
-		if ctxErr := subCtx.Err(); ctxErr != nil {
-			c.debug.Println(fmt.Sprintf("terminated due to context: %v", ctxErr))
-			return nil, ctxErr
+	c.workers.Add(1)
+	go func() {
+		defer c.MIDs.Free(sp.PacketID)
+		defer c.workers.Done()
+		defer close(resp)
+
+		c.debug.Println("sending SUBSCRIBE for %d", sp.PacketID)
+		if _, err := sp.WriteTo(c.Conn); err != nil {
+			c.errors.Printf("failed to send SUBSCRIBE for %d: %w", sp.PacketID, err)
+			return
 		}
-	case sap = <-cpCtx.Return:
-	}
 
-	if sap.Type != packets.SUBACK {
-		return nil, fmt.Errorf("received %d instead of Suback", sap.Type)
-	}
-	c.debug.Println("received SUBACK")
+		var receivedPacket packets.ControlPacket
 
-	sa := SubackFromPacketSuback(sap.Content.(*packets.Suback))
-	switch {
-	case len(sa.Reasons) == 1:
-		if sa.Reasons[0] >= 0x80 {
-			var reason string
-			c.debug.Println("received an error code in Suback:", sa.Reasons[0])
-			if sa.Properties != nil {
-				reason = sa.Properties.ReasonString
-			}
-			return sa, fmt.Errorf("failed to subscribe to topic: %s", reason)
-		}
```

Comment:
In my implementation, I do not return an error if the server returns an error reason code. I made the same change for Unsubscribe() and Publish()

IMO, it should not be considered an error from the client perspective, it is just one of the possible results from a successfully completed operation. Curious to hear thoughts on this.